File: java/src/main/java/io/github/emindeniz99/applepurchase/receipt/ReceiptVerifier.java
package io.github.emindeniz99.applepurchase.receipt;

import io.github.emindeniz99.applepurchase.VerificationException;
import io.github.emindeniz99.applepurchase.VerificationException.Reason;
import org.bouncycastle.asn1.ASN1Encodable;
import org.bouncycastle.asn1.ASN1Integer;
import org.bouncycastle.asn1.ASN1OctetString;
import org.bouncycastle.asn1.ASN1Primitive;
import org.bouncycastle.asn1.ASN1Sequence;
import org.bouncycastle.asn1.ASN1Set;
import org.bouncycastle.asn1.ASN1String;
import org.bouncycastle.cert.X509CertificateHolder;
import org.bouncycastle.cert.jcajce.JcaX509CertificateConverter;
import org.bouncycastle.cms.CMSException;
import org.bouncycastle.cms.CMSSignedData;
import org.bouncycastle.cms.SignerInformation;
import org.bouncycastle.cms.jcajce.JcaSimpleSignerInfoVerifierBuilder;
import org.bouncycastle.jce.provider.BouncyCastleProvider;
import org.bouncycastle.operator.OperatorCreationException;

import java.io.IOException;
import java.math.BigInteger;
import java.security.GeneralSecurityException;
import java.security.MessageDigest;
import java.security.cert.CertPathBuilder;
import java.security.cert.CertPathBuilderException;
import java.security.cert.CertStore;
import java.security.cert.CollectionCertStoreParameters;
import java.security.cert.PKIXBuilderParameters;
import java.security.cert.TrustAnchor;
import java.security.cert.X509CertSelector;
import java.security.cert.X509Certificate;
import java.time.Instant;
import java.time.format.DateTimeParseException;
import java.util.ArrayList;
import java.util.Base64;
import java.util.Collection;
import java.util.Date;
import java.util.HashSet;
import java.util.Iterator;
import java.util.List;
import java.util.Set;

/**
 * Verifies legacy PKCS#7 app receipts (the blob apps used to send to the
 * deprecated {@code verifyReceipt} endpoint) completely offline, against the
 * pinned Apple Inc. Root CA — the server-side port of Apple's "Validating
 * receipts on the device" procedure (PLAN.md §2.2).
 *
 * <p>Thread-safe once constructed.</p>
 */
public final class ReceiptVerifier {

    // Receipt attribute types (Apple, "Validating receipts on the device").
    private static final int ATTR_BUNDLE_ID = 2;
    private static final int ATTR_APP_VERSION = 3;
    private static final int ATTR_OPAQUE_VALUE = 4;
    private static final int ATTR_SHA1_HASH = 5;
    private static final int ATTR_CREATION_DATE = 12;
    private static final int ATTR_IN_APP = 17;
    private static final int ATTR_ORIGINAL_APP_VERSION = 19;
    private static final int ATTR_EXPIRATION_DATE = 21;

    private static final int IAP_QUANTITY = 1701;
    private static final int IAP_PRODUCT_ID = 1702;
    private static final int IAP_TRANSACTION_ID = 1703;
    private static final int IAP_PURCHASE_DATE = 1704;
    private static final int IAP_ORIGINAL_TRANSACTION_ID = 1705;
    private static final int IAP_ORIGINAL_PURCHASE_DATE = 1706;
    private static final int IAP_EXPIRES_DATE = 1708;
    private static final int IAP_WEB_ORDER_LINE_ITEM_ID = 1711;
    private static final int IAP_CANCELLATION_DATE = 1712;
    private static final int IAP_IS_IN_INTRO_OFFER_PERIOD = 1719;

    private final Set<TrustAnchor> trustAnchors;
    private final String bundleId;
    private final BouncyCastleProvider provider = new BouncyCastleProvider();

    /**
     * @param trustedRoots pinned root CAs (production:
     *                     {@code AppleRootCerts.receiptRoots()})
     * @param bundleId     the app's bundle id the receipt must carry
     */
    public ReceiptVerifier(Set<X509Certificate> trustedRoots, String bundleId) {
        if (trustedRoots == null || trustedRoots.isEmpty()) {
            throw new IllegalArgumentException("trustedRoots must not be empty");
        }
        if (bundleId == null) {
            throw new IllegalArgumentException("bundleId must not be null");
        }
        Set<TrustAnchor> anchors = new HashSet<TrustAnchor>();
        for (X509Certificate root : trustedRoots) {
            anchors.add(new TrustAnchor(root, null));
        }
        this.trustAnchors = anchors;
        this.bundleId = bundleId;
    }

    /** Verifies a base64-encoded receipt (the usual client transport form). */
    public AppReceipt verify(String base64Receipt) throws VerificationException {
        if (base64Receipt == null) {
            throw new VerificationException(Reason.INVALID_RECEIPT_FORMAT, "receipt is null");
        }
        byte[] der;
        try {
            der = Base64.getMimeDecoder().decode(base64Receipt);
        } catch (IllegalArgumentException e) {
            throw new VerificationException(Reason.INVALID_RECEIPT_FORMAT, "receipt is not valid base64", e);
        }
        return verify(der);
    }

    /** Verifies a DER-encoded PKCS#7 receipt. */
    public AppReceipt verify(byte[] receiptDer) throws VerificationException {
        return verify(receiptDer, null);
    }

    /**
     * Verifies a receipt and additionally enforces the device-hash binding:
     * {@code SHA1(deviceGuid ‖ opaqueValue ‖ bundleIdBytes)} must equal
     * attribute 5. Optional because it requires the client to send its
     * {@code identifierForVendor} bytes (PLAN.md D4) — each device's own
     * receipt embeds that device's GUID, so cross-device restore still works:
     * every device presents its own receipt.
     */
    public AppReceipt verify(byte[] receiptDer, byte[] deviceGuid) throws VerificationException {
        if (receiptDer == null) {
            throw new VerificationException(Reason.INVALID_RECEIPT_FORMAT, "receipt is null");
        }
        CMSSignedData cms;
        try {
            cms = new CMSSignedData(receiptDer);
        } catch (CMSException e) {
            throw new VerificationException(Reason.INVALID_RECEIPT_FORMAT, "not a PKCS#7/CMS blob", e);
        }
        if (cms.getSignedContent() == null || !(cms.getSignedContent().getContent() instanceof byte[])) {
            throw new VerificationException(Reason.INVALID_RECEIPT_FORMAT, "no encapsulated payload");
        }
        byte[] payload = (byte[]) cms.getSignedContent().getContent();

        // Parsed before signature verification only to learn the creation
        // date (chain validity is anchored at signing time); nothing from it
        // is trusted until after the chain + signature checks pass.
        AppReceipt receipt = parsePayload(payload);
        Date at = receipt.creationDate() != null ? Date.from(receipt.creationDate()) : new Date();

        X509Certificate signerCert = validateChain(cms, at);
        verifyCmsSignature(cms, signerCert);

        if (!bundleId.equals(receipt.bundleId())) {
            throw new VerificationException(Reason.WRONG_BUNDLE_ID,
                    "expected " + bundleId + " but receipt has " + receipt.bundleId());
        }
        if (deviceGuid != null) {
            verifyDeviceHash(receipt, deviceGuid);
        }
        return receipt;
    }

    /** PKIX-builds signer → (intermediates from the CMS) → pinned root at {@code at}. */
    private X509Certificate validateChain(CMSSignedData cms, Date at) throws VerificationException {
        Iterator<SignerInformation> signers = cms.getSignerInfos().getSigners().iterator();
        if (!signers.hasNext()) {
            throw new VerificationException(Reason.INVALID_RECEIPT_FORMAT, "no signer info");
        }
        SignerInformation signer = signers.next();
        Collection<X509CertificateHolder> matches = cms.getCertificates().getMatches(signer.getSID());
        if (matches.isEmpty()) {
            throw new VerificationException(Reason.INVALID_RECEIPT_FORMAT, "signer certificate not embedded");
        }
        try {
            JcaX509CertificateConverter converter = new JcaX509CertificateConverter();
            X509Certificate signerCert = converter.getCertificate(matches.iterator().next());
            List<X509Certificate> embedded = new ArrayList<X509Certificate>();
            for (X509CertificateHolder holder : cms.getCertificates().getMatches(null)) {
                embedded.add(converter.getCertificate(holder));
            }
            X509CertSelector target = new X509CertSelector();
            target.setCertificate(signerCert);
            PKIXBuilderParameters params = new PKIXBuilderParameters(trustAnchors, target);
            params.addCertStore(CertStore.getInstance("Collection",
                    new CollectionCertStoreParameters(embedded)));
            params.setRevocationEnabled(false);
            params.setDate(at);
            CertPathBuilder.getInstance("PKIX").build(params);
            return signerCert;
        } catch (CertPathBuilderException e) {
            throw new VerificationException(Reason.INVALID_CHAIN,
                    "signer chain does not validate to a pinned Apple root: " + e.getMessage(), e);
        } catch (GeneralSecurityException e) {
            throw new VerificationException(Reason.INVALID_CHAIN, "chain validation unavailable", e);
        }
    }

    private void verifyCmsSignature(CMSSignedData cms, X509Certificate signerCert)
            throws VerificationException {
        try {
            SignerInformation signer = cms.getSignerInfos().getSigners().iterator().next();
            boolean valid = signer.verify(new JcaSimpleSignerInfoVerifierBuilder()
                    .setProvider(provider)
                    .build(signerCert));
            if (!valid) {
                throw new VerificationException(Reason.INVALID_SIGNATURE, "CMS signature check failed");
            }
        } catch (CMSException e) {
            throw new VerificationException(Reason.INVALID_SIGNATURE, "CMS signature check failed", e);
        } catch (OperatorCreationException e) {
            throw new VerificationException(Reason.INVALID_SIGNATURE, "CMS signature check errored", e);
        }
    }

    private static void verifyDeviceHash(AppReceipt receipt, byte[] deviceGuid)
            throws VerificationException {
        if (receipt.opaqueValue() == null || receipt.sha1Hash() == null
                || receipt.bundleIdBytes() == null) {
            throw new VerificationException(Reason.DEVICE_HASH_MISMATCH,
                    "receipt lacks the attributes needed for the device-hash check");
        }
        try {
            MessageDigest sha1 = MessageDigest.getInstance("SHA-1");
            sha1.update(deviceGuid);
            sha1.update(receipt.opaqueValue());
            sha1.update(receipt.bundleIdBytes());
            if (!MessageDigest.isEqual(sha1.digest(), receipt.sha1Hash())) {
                throw new VerificationException(Reason.DEVICE_HASH_MISMATCH,
                        "computed device hash does not match attribute 5");
            }
        } catch (GeneralSecurityException e) {
            throw new VerificationException(Reason.DEVICE_HASH_MISMATCH, "SHA-1 unavailable", e);
        }
    }

    // --- ASN.1 payload parsing -------------------------------------------

    private static AppReceipt parsePayload(byte[] payload) throws VerificationException {
        ASN1Set attributes = parseAttributeSet(payload, "receipt payload");
        String parsedBundleId = null;
        byte[] bundleIdBytes = null;
        String appVersion = null;
        byte[] opaqueValue = null;
        byte[] sha1Hash = null;
        Instant creationDate = null;
        String originalAppVersion = null;
        Instant expirationDate = null;
        List<InAppPurchase> purchases = new ArrayList<InAppPurchase>();

        for (ASN1Encodable element : attributes) {
            Attribute attr = Attribute.of(element);
            switch (attr.type) {
                case ATTR_BUNDLE_ID:
                    parsedBundleId = decodeString(attr.value);
                    bundleIdBytes = attr.value;
                    break;
                case ATTR_APP_VERSION:
                    appVersion = decodeString(attr.value);
                    break;
                case ATTR_OPAQUE_VALUE:
                    opaqueValue = attr.value;
                    break;
                case ATTR_SHA1_HASH:
                    sha1Hash = attr.value;
                    break;
                case ATTR_CREATION_DATE:
                    creationDate = decodeDate(attr.value);
                    break;
                case ATTR_IN_APP:
                    purchases.add(parseInApp(attr.value));
                    break;
                case ATTR_ORIGINAL_APP_VERSION:
                    originalAppVersion = decodeString(attr.value);
                    break;
                case ATTR_EXPIRATION_DATE:
                    expirationDate = decodeDate(attr.value);
                    break;
                default:
                    // Receipts carry undocumented attribute types; ignore them.
                    break;
            }
        }
        return new AppReceipt(parsedBundleId, bundleIdBytes, appVersion, opaqueValue,
                sha1Hash, creationDate, originalAppVersion, expirationDate, purchases);
    }

    private static InAppPurchase parseInApp(byte[] inAppSet) throws VerificationException {
        ASN1Set attributes = parseAttributeSet(inAppSet, "in-app purchase attribute");
        Long quantity = null;
        String productId = null;
        String transactionId = null;
        String originalTransactionId = null;
        Instant purchaseDate = null;
        Instant originalPurchaseDate = null;
        Instant expiresDate = null;
        Instant cancellationDate = null;
        Long webOrderLineItemId = null;
        Long isInIntroOfferPeriod = null;

        for (ASN1Encodable element : attributes) {
            Attribute attr = Attribute.of(element);
            switch (attr.type) {
                case IAP_QUANTITY:
                    quantity = decodeInteger(attr.value);
                    break;
                case IAP_PRODUCT_ID:
                    productId = decodeString(attr.value);
                    break;
                case IAP_TRANSACTION_ID:
                    transactionId = decodeString(attr.value);
                    break;
                case IAP_PURCHASE_DATE:
                    purchaseDate = decodeDate(attr.value);
                    break;
                case IAP_ORIGINAL_TRANSACTION_ID:
                    originalTransactionId = decodeString(attr.value);
                    break;
                case IAP_ORIGINAL_PURCHASE_DATE:
                    originalPurchaseDate = decodeDate(attr.value);
                    break;
                case IAP_EXPIRES_DATE:
                    expiresDate = decodeDate(attr.value);
                    break;
                case IAP_WEB_ORDER_LINE_ITEM_ID:
                    webOrderLineItemId = decodeInteger(attr.value);
                    break;
                case IAP_CANCELLATION_DATE:
                    cancellationDate = decodeDate(attr.value);
                    break;
                case IAP_IS_IN_INTRO_OFFER_PERIOD:
                    isInIntroOfferPeriod = decodeInteger(attr.value);
                    break;
                default:
                    break;
            }
        }
        return new InAppPurchase(quantity, productId, transactionId, originalTransactionId,
                purchaseDate, originalPurchaseDate, expiresDate, cancellationDate,
                webOrderLineItemId, isInIntroOfferPeriod);
    }

    private static ASN1Set parseAttributeSet(byte[] der, String what) throws VerificationException {
        ASN1Primitive parsed;
        try {
            parsed = ASN1Primitive.fromByteArray(der);
        } catch (IOException e) {
            throw new VerificationException(Reason.INVALID_RECEIPT_FORMAT, what + " is not valid ASN.1", e);
        }
        if (!(parsed instanceof ASN1Set)) {
            throw new VerificationException(Reason.INVALID_RECEIPT_FORMAT, what + " is not an ASN.1 SET");
        }
        return (ASN1Set) parsed;
    }

    /** {@code ReceiptAttribute ::= SEQUENCE { type INTEGER, version INTEGER, value OCTET STRING }} */
    private static final class Attribute {
        final int type;
        final byte[] value;

        private Attribute(int type, byte[] value) {
            this.type = type;
            this.value = value;
        }

        static Attribute of(ASN1Encodable element) throws VerificationException {
            try {
                ASN1Sequence seq = ASN1Sequence.getInstance(element);
                if (seq.size() < 3) {
                    throw new VerificationException(Reason.INVALID_RECEIPT_FORMAT,
                            "receipt attribute has " + seq.size() + " fields, expected 3");
                }
                BigInteger type = ASN1Integer.getInstance(seq.getObjectAt(0)).getValue();
                byte[] value = ASN1OctetString.getInstance(seq.getObjectAt(2)).getOctets();
                return new Attribute(type.intValueExact(), value);
            } catch (IllegalArgumentException e) {
                throw new VerificationException(Reason.INVALID_RECEIPT_FORMAT,
                        "malformed receipt attribute", e);
            } catch (ArithmeticException e) {
                throw new VerificationException(Reason.INVALID_RECEIPT_FORMAT,
                        "receipt attribute type out of range", e);
            }
        }
    }

    private static String decodeString(byte[] der) throws VerificationException {
        try {
            ASN1Primitive parsed = ASN1Primitive.fromByteArray(der);
            if (!(parsed instanceof ASN1String)) {
                throw new VerificationException(Reason.INVALID_RECEIPT_FORMAT,
                        "attribute value is not an ASN.1 string");
            }
            return ((ASN1String) parsed).getString();
        } catch (IOException e) {
            throw new VerificationException(Reason.INVALID_RECEIPT_FORMAT,
                    "attribute value is not valid ASN.1", e);
        }
    }

    private static Long decodeInteger(byte[] der) throws VerificationException {
        try {
            ASN1Primitive parsed = ASN1Primitive.fromByteArray(der);
            if (!(parsed instanceof ASN1Integer)) {
                throw new VerificationException(Reason.INVALID_RECEIPT_FORMAT,
                        "attribute value is not an ASN.1 integer");
            }
            return ((ASN1Integer) parsed).getValue().longValueExact();
        } catch (IOException e) {
            throw new VerificationException(Reason.INVALID_RECEIPT_FORMAT,
                    "attribute value is not valid ASN.1", e);
        } catch (ArithmeticException e) {
            throw new VerificationException(Reason.INVALID_RECEIPT_FORMAT,
                    "attribute integer out of range", e);
        }
    }

    /** RFC 3339 date in an IA5String; empty means absent (real receipts do this). */
    private static Instant decodeDate(byte[] der) throws VerificationException {
        String text = decodeString(der);
        if (text.isEmpty()) {
            return null;
        }
        try {
            return Instant.parse(text);
        } catch (DateTimeParseException e) {
            throw new VerificationException(Reason.INVALID_RECEIPT_FORMAT,
                    "unparseable receipt date: " + text, e);
        }
    }
}

File: swift/Sources/ApplePurchaseVerifier/ReceiptVerifier.swift
import Crypto
import _CryptoExtras
import Foundation
import SwiftASN1
import X509

/// One in-app purchase from a legacy app receipt (attribute 17).
public struct InAppPurchase: Sendable {
    public var quantity: Int64?
    public var productId: String?
    public var transactionId: String?
    public var originalTransactionId: String?
    public var purchaseDate: Date?
    public var originalPurchaseDate: Date?
    public var expiresDate: Date?
    public var cancellationDate: Date?
    public var webOrderLineItemId: Int64?
    public var isInIntroOfferPeriod: Int64?
}

/// A verified legacy app receipt. Only receipts returned by
/// ``ReceiptVerifier`` should be trusted.
public struct AppReceipt: Sendable {
    public var bundleId: String?
    public var bundleIdBytes: Data?
    public var appVersion: String?
    public var opaqueValue: Data?
    public var sha1Hash: Data?
    public var creationDate: Date?
    public var originalAppVersion: String?
    public var expirationDate: Date?
    public var inAppPurchases: [InAppPurchase] = []
}

/// Verifies legacy PKCS#7 app receipts completely offline against the
/// pinned Apple Inc. Root CA — the server-side port of Apple's "Validating
/// receipts on the device" procedure (PLAN.md §2.2), mirroring the Java
/// implementation. CMS parsing accepts BER (genuine Apple/Xcode receipts
/// use indefinite lengths).
public struct ReceiptVerifier: Sendable {
    private let roots: [Certificate]
    private let bundleId: String

    /// - Parameters:
    ///   - trustedRoots: pinned DER roots (production: ``appleReceiptRoots()``)
    ///   - bundleId: bundle id the receipt must carry
    public init(trustedRoots: [Data], bundleId: String) throws {
        guard !trustedRoots.isEmpty, !bundleId.isEmpty else {
            throw VerificationError(.invalidCertificate, "trustedRoots and bundleId are required")
        }
        self.roots = try trustedRoots.map { try Certificate(derEncoded: [UInt8]($0)) }
        self.bundleId = bundleId
    }

    /// Verifies a base64 receipt (the usual client transport form).
    public func verify(base64Receipt: String, deviceGuid: Data? = nil) async throws -> AppReceipt {
        guard let der = Data(base64Encoded: base64Receipt,
                             options: [.ignoreUnknownCharacters]) else {
            throw VerificationError(.invalidReceiptFormat, "receipt is not valid base64")
        }
        return try await verify(receipt: der, deviceGuid: deviceGuid)
    }

    /// Verifies a DER/BER receipt. Passing `deviceGuid` additionally
    /// enforces the device-hash binding: SHA1(guid ‖ opaqueValue ‖
    /// bundleIdBytes) must equal attribute 5 (optional — PLAN.md D4).
    public func verify(receipt: Data, deviceGuid: Data? = nil) async throws -> AppReceipt {
        let cms = try CMSReceipt(parsing: [UInt8](receipt))

        // Parsed before signature verification only to learn the creation
        // date (chain validity anchors at signing time); nothing from it is
        // trusted until the chain + signature checks pass.
        let fields = try parsePayload(cms.content)
        let at = fields.creationDate ?? Date()

        let signerCert = try cms.signerCertificate()
        var verifier = Verifier(rootCertificates: CertificateStore(roots)) {
            RFC5280Policy(validationTime: at)
        }
        let result = await verifier.validate(
            leafCertificate: signerCert,
            intermediates: CertificateStore(cms.certificates))
        if case .couldNotValidate = result {
            throw VerificationError(.invalidChain,
                "signer chain does not validate to a pinned root")
        }
        try cms.verifySignature(signerCert: signerCert)

        guard fields.bundleId == bundleId else {
            throw VerificationError(.wrongBundleId,
                "expected \(bundleId) but receipt has \(fields.bundleId ?? "nil")")
        }
        if let deviceGuid {
            try verifyDeviceHash(fields, deviceGuid: deviceGuid)
        }
        return fields
    }

    private func verifyDeviceHash(_ fields: AppReceipt, deviceGuid: Data) throws {
        guard let opaque = fields.opaqueValue, let expected = fields.sha1Hash,
              let bundleBytes = fields.bundleIdBytes else {
            throw VerificationError(.deviceHashMismatch,
                "receipt lacks the attributes needed for the device-hash check")
        }
        var input = Data()
        input.append(deviceGuid)
        input.append(opaque)
        input.append(bundleBytes)
        let computed = Data(Insecure.SHA1.hash(data: input))
        // Constant-time comparison.
        guard computed.count == expected.count,
              zip(computed, expected).reduce(0, { $0 | ($1.0 ^ $1.1) }) == 0 else {
            throw VerificationError(.deviceHashMismatch,
                "computed device hash does not match attribute 5")
        }
    }
}

// MARK: - CMS SignedData (BER-tolerant)

private struct CMSReceipt {
    let content: [UInt8]
    let certificates: [Certificate]
    let certificateNodes: [(serial: [UInt8], issuer: [UInt8])]
    let signerIssuer: [UInt8]
    let signerSerial: [UInt8]
    let digestName: String
    let signedAttrsBytes: [UInt8]?
    let messageDigest: [UInt8]?
    let signature: [UInt8]

    init(parsing bytes: [UInt8]) throws {
        let root: ASN1Node
        do {
            root = try BER.parse(bytes)
        } catch {
            throw VerificationError(.invalidReceiptFormat, "not parseable ASN.1")
        }
        do {
            let contentInfo = try Self.children(root)
            let contentType = try ASN1ObjectIdentifier(derEncoded: contentInfo[0])
            guard contentType == [1, 2, 840, 113549, 1, 7, 2] else {
                throw VerificationError(.invalidReceiptFormat, "not CMS SignedData")
            }
            let signedData = try Self.children(try Self.explicit(contentInfo[1]))

            // encapContentInfo: SEQ { OID, [0] EXPLICIT OCTET STRING }
            let encap = try Self.children(signedData[2])
            guard encap.count >= 2 else {
                throw VerificationError(.invalidReceiptFormat, "no encapsulated payload")
            }
            self.content = try Self.octetStringValue(try Self.explicit(encap[1]))

            var certificates: [Certificate] = []
            var certificateNodes: [(serial: [UInt8], issuer: [UInt8])] = []
            for node in signedData.dropFirst(3) where node.identifier.tagClass == .contextSpecific
                && node.identifier.tagNumber == 0 {
                for certNode in try Self.children(node) {
                    let der = [UInt8](certNode.encodedBytes)
                    certificates.append(try Certificate(derEncoded: der))
                    let tbs = try Self.children(try Self.children(DER.parse(der))[0])
                    var index = 0
                    if tbs[0].identifier.tagClass == .contextSpecific { index = 1 }
                    certificateNodes.append((serial: try Self.primitive(tbs[index]),
                                             issuer: [UInt8](tbs[index + 2].encodedBytes)))
                }
            }
            self.certificates = certificates
            self.certificateNodes = certificateNodes

            guard let signerInfos = signedData.last,
                  signerInfos.identifier == .set,
                  let signerInfo = try Self.children(signerInfos).first else {
                throw VerificationError(.invalidReceiptFormat, "no signer info")
            }
            let signerFields = try Self.children(signerInfo)
            let sid = try Self.children(signerFields[1])
            self.signerIssuer = [UInt8](sid[0].encodedBytes)
            self.signerSerial = try Self.primitive(sid[1])

            let digestAlgOid = try ASN1ObjectIdentifier(
                derEncoded: try Self.children(signerFields[2])[0])
            switch digestAlgOid {
            case [1, 3, 14, 3, 2, 26]: self.digestName = "sha1"
            case [2, 16, 840, 1, 101, 3, 4, 2, 1]: self.digestName = "sha256"
            default:
                throw VerificationError(.invalidReceiptFormat, "unsupported digest algorithm")
            }

            var index = 3
            if signerFields[index].identifier.tagClass == .contextSpecific
                && signerFields[index].identifier.tagNumber == 0 {
                let attrsNode = signerFields[index]
                // Signature covers the signedAttrs re-encoded as an explicit
                // SET (RFC 5652 §5.4): swap the IMPLICIT [0] tag for SET.
                var reencoded = [UInt8](attrsNode.encodedBytes)
                reencoded[0] = 0x31
                self.signedAttrsBytes = reencoded
                var digest: [UInt8]? = nil
                for attr in try Self.children(attrsNode) {
                    let parts = try Self.children(attr)
                    if try ASN1ObjectIdentifier(derEncoded: parts[0])
                        == [1, 2, 840, 113549, 1, 9, 4] {
                        digest = try Self.primitive(try Self.children(parts[1])[0])
                    }
                }
                self.messageDigest = digest
                index += 1
            } else {
                self.signedAttrsBytes = nil
                self.messageDigest = nil
            }
            index += 1 // signatureAlgorithm — RSA PKCS#1 v1.5, digest drives the hash
            self.signature = try Self.primitive(signerFields[index])
        } catch let error as VerificationError {
            throw error
        } catch {
            throw VerificationError(.invalidReceiptFormat, "malformed CMS structure")
        }
    }

    func signerCertificate() throws -> Certificate {
        for (offset, node) in certificateNodes.enumerated()
        where node.serial == signerSerial && node.issuer == signerIssuer {
            return certificates[offset]
        }
        throw VerificationError(.invalidReceiptFormat, "signer certificate not embedded")
    }

    func verifySignature(signerCert: Certificate) throws {
        guard let publicKey = _RSA.Signing.PublicKey(signerCert.publicKey) else {
            throw VerificationError(.invalidSignature, "signer key is not RSA")
        }
        let signedBytes: [UInt8]
        if let signedAttrsBytes {
            let contentDigest: [UInt8] = digestName == "sha1"
                ? [UInt8](Insecure.SHA1.hash(data: Data(content)))
                : [UInt8](SHA256.hash(data: Data(content)))
            guard let messageDigest, messageDigest == contentDigest else {
                throw VerificationError(.invalidSignature,
                    "messageDigest attribute does not match content")
            }
            signedBytes = signedAttrsBytes
        } else {
            signedBytes = content
        }
        let rsaSignature = _RSA.Signing.RSASignature(rawRepresentation: Data(signature))
        let valid: Bool
        if digestName == "sha1" {
            valid = publicKey.isValidSignature(rsaSignature,
                for: Insecure.SHA1.hash(data: Data(signedBytes)), padding: .insecurePKCS1v1_5)
        } else {
            valid = publicKey.isValidSignature(rsaSignature,
                for: SHA256.hash(data: Data(signedBytes)), padding: .insecurePKCS1v1_5)
        }
        guard valid else {
            throw VerificationError(.invalidSignature, "CMS signature check failed")
        }
    }

    static func children(_ node: ASN1Node) throws -> [ASN1Node] {
        guard case .constructed(let nodes) = node.content else {
            throw VerificationError(.invalidReceiptFormat, "expected constructed ASN.1 node")
        }
        return Array(nodes)
    }

    /// Unwraps an EXPLICIT context tag ([0] { inner }) to its inner node.
    static func explicit(_ node: ASN1Node) throws -> ASN1Node {
        let inner = try children(node)
        guard inner.count == 1 else {
            throw VerificationError(.invalidReceiptFormat, "expected single explicit content")
        }
        return inner[0]
    }

    static func primitive(_ node: ASN1Node) throws -> [UInt8] {
        guard case .primitive(let bytes) = node.content else {
            throw VerificationError(.invalidReceiptFormat, "expected primitive ASN.1 node")
        }
        return [UInt8](bytes)
    }

    /// Value bytes of an OCTET STRING, joining BER constructed chunks.
    static func octetStringValue(_ node: ASN1Node) throws -> [UInt8] {
        switch node.content {
        case .primitive(let bytes):
            return [UInt8](bytes)
        case .constructed(let chunks):
            var out: [UInt8] = []
            for chunk in chunks {
                out.append(contentsOf: try octetStringValue(chunk))
            }
            return out
        }
    }
}

// MARK: - Receipt payload (strict DER)

private let attrBundleId = 2
private let attrAppVersion = 3
private let attrOpaqueValue = 4
private let attrSha1Hash = 5
private let attrCreationDate = 12
private let attrInApp = 17
private let attrOriginalAppVersion = 19
private let attrExpirationDate = 21

private func parsePayload(_ content: [UInt8]) throws -> AppReceipt {
    var receipt = AppReceipt()
    for (type, value) in try parseAttributeSet(content) {
        switch type {
        case attrBundleId:
            receipt.bundleId = try decodeString(value)
            receipt.bundleIdBytes = Data(value)
        case attrAppVersion: receipt.appVersion = try decodeString(value)
        case attrOpaqueValue: receipt.opaqueValue = Data(value)
        case attrSha1Hash: receipt.sha1Hash = Data(value)
        case attrCreationDate: receipt.creationDate = try decodeDate(value)
        case attrInApp: receipt.inAppPurchases.append(try parseInApp(value))
        case attrOriginalAppVersion: receipt.originalAppVersion = try decodeString(value)
        case attrExpirationDate: receipt.expirationDate = try decodeDate(value)
        default: break // receipts carry undocumented attribute types
        }
    }
    return receipt
}

private func parseInApp(_ value: [UInt8]) throws -> InAppPurchase {
    var purchase = InAppPurchase()
    for (type, v) in try parseAttributeSet(value) {
        switch type {
        case 1701: purchase.quantity = try decodeInteger(v)
        case 1702: purchase.productId = try decodeString(v)
        case 1703: purchase.transactionId = try decodeString(v)
        case 1704: purchase.purchaseDate = try decodeDate(v)
        case 1705: purchase.originalTransactionId = try decodeString(v)
        case 1706: purchase.originalPurchaseDate = try decodeDate(v)
        case 1708: purchase.expiresDate = try decodeDate(v)
        case 1711: purchase.webOrderLineItemId = try decodeInteger(v)
        case 1712: purchase.cancellationDate = try decodeDate(v)
        case 1719: purchase.isInIntroOfferPeriod = try decodeInteger(v)
        default: break
        }
    }
    return purchase
}

private func parseAttributeSet(_ der: [UInt8]) throws -> [(Int, [UInt8])] {
    let root: ASN1Node
    do {
        root = try DER.parse(der)
    } catch {
        throw VerificationError(.invalidReceiptFormat, "attribute set is not valid ASN.1")
    }
    guard root.identifier == .set else {
        throw VerificationError(.invalidReceiptFormat, "attribute set is not an ASN.1 SET")
    }
    var attributes: [(Int, [UInt8])] = []
    for child in try CMSReceipt.children(root) {
        let fields = try CMSReceipt.children(child)
        guard child.identifier == .sequence, fields.count >= 3,
              fields[0].identifier == .integer, fields[2].identifier == .octetString else {
            throw VerificationError(.invalidReceiptFormat, "malformed receipt attribute")
        }
        attributes.append((try intValue(try CMSReceipt.primitive(fields[0])),
                           try CMSReceipt.primitive(fields[2])))
    }
    return attributes
}

private func intValue(_ contents: [UInt8]) throws -> Int {
    guard contents.count <= 6 else {
        throw VerificationError(.invalidReceiptFormat, "attribute integer out of range")
    }
    if let first = contents.first, first >= 0x80 {
        throw VerificationError(.invalidReceiptFormat, "negative receipt integer")
    }
    return contents.reduce(0) { $0 * 256 + Int($1) }
}

private func parseNested(_ der: [UInt8]) throws -> ASN1Node {
    do {
        return try DER.parse(der)
    } catch {
        throw VerificationError(.invalidReceiptFormat, "attribute value is not valid ASN.1")
    }
}

private func decodeString(_ der: [UInt8]) throws -> String {
    let node = try parseNested(der)
    guard node.identifier == .utf8String || node.identifier == .ia5String,
          let text = String(bytes: try CMSReceipt.primitive(node), encoding: .utf8) else {
        throw VerificationError(.invalidReceiptFormat, "attribute value is not an ASN.1 string")
    }
    return text
}

private func decodeInteger(_ der: [UInt8]) throws -> Int64 {
    let node = try parseNested(der)
    guard node.identifier == .integer else {
        throw VerificationError(.invalidReceiptFormat, "attribute value is not an ASN.1 integer")
    }
    return Int64(try intValue(try CMSReceipt.primitive(node)))
}

/// RFC 3339 date in an IA5String; empty means absent (real receipts do this).
private func decodeDate(_ der: [UInt8]) throws -> Date? {
    let text = try decodeString(der)
    if text.isEmpty { return nil }
    let formatter = ISO8601DateFormatter()
    formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
    if let date = formatter.date(from: text) { return date }
    formatter.formatOptions = [.withInternetDateTime]
    guard let date = formatter.date(from: text) else {
        throw VerificationError(.invalidReceiptFormat, "unparseable receipt date: \(text)")
    }
    return date
}
